pyboy/utils.py

```diff
@@ -273,3 +274,146 @@ def __init__(
+        
+        for bcd_byte in bcd_bytes:
+            decimal_value += ((bcd_byte >> 4) * 10 + (bcd_byte & 0x0F)) * self._multiplier
+            self._multiplier *= 100
+
+        return decimal_value
+
+class MemoryScanner():
+    """A class for scanning memory within a given range."""
+
+    def __init__(self, pyboy):
+        """
+        Initializes the MemoryScanner with a pyboy instance.
+
+        :param pyboy: The pyboy emulator instance.
+        """
+        self.pyboy = pyboy
+        self.bcd_converter = BCDConverter()
+
+    def read_memory(self, address, byte_width=1, value_type=ScanMode.INT, endian_type=EndianType.LITTLE):
+        """
+        Reads memory at the specified address.
+
+        :param address: The address to read.
+        :param byte_width: The number of bytes to consider for each value.
+        :param value_type: The type of value (INT or BCD) to consider.
+        :param endian_type: The endian type to use. Note, this is only used for 16-bit values and higher.
+        :return: The value at the specified address.
+        """
+        value_bytes = [self.pyboy.get_memory_value(address + i) for i in range(byte_width)]
+        value = int.from_bytes(value_bytes, 'little' if endian_type == EndianType.LITTLE else 'big')
+
+        if value_type == ScanMode.BCD:
+            value = self.bcd_converter.bcd_to_dec(value, byte_width, endian_type)
+
+        return value
+
+    def scan_memory(self, start_addr, end_addr, target_value, compare_type=CompareType.EXACT, value_type=ScanMode.INT, byte_width=1, endian_type=EndianType.LITTLE):
+        """
+        Scans memory in the specified range, looking for a target value.
+
+        :param start_addr: The starting address for the scan.
+        :param end_addr: The ending address for the scan.
+        :param target_value: The value to search for.
+        :param compare_type: The type of comparison to use.
+        :param value_type: The type of value (INT or BCD) to consider.
+        :param byte_width: The number of bytes to consider for each value.
+        :param endian_type: The endian type to use. Note, this is only used for 16-bit values and higher.
+        :return: A list of addresses where the target value is found.
+        """
+        found_addresses = []
+        for addr in range(start_addr, end_addr - byte_width + 2): # Adjust the loop to prevent reading past end_addr
+            # Read multiple bytes based on byte_width and endian_type
+            value_bytes = [self.pyboy.get_memory_value(addr + i) for i in range(byte_width)]
+            value = int.from_bytes(value_bytes, 'little' if endian_type == EndianType.LITTLE else 'big')
+
+            if value_type == ScanMode.BCD:
+                value = self.bcd_converter.bcd_to_dec(value, byte_width, endian_type)
+
+            if self._check_value(value, target_value, compare_type.value):
+                found_addresses.append(addr)
+
+        return found_addresses
```

Comment:
Many years ago, I used Cheat Engine. It had the option to narrow down memory addresses, without knowing the value, by making an initial request like "all value under 10", then run the game a bit and then "of all those values, find the ones that increased". Would this be possible to incorporate?

So more or less the same as now, but with the iterative/recursive searching.

Reply:
Take a look at memoryscanner.py when you get a chance, I implemented a rescan method that has more open ended comparisons like all values that increased. 

I did add one comparison method called MATCH where they can pass an optional param new_value, but i couldn't make new_value default to None because uint64_t cant be null. So the default value is INT64_MAX. Let me know your thoughts on this implementation.

Reply:
However i am also concerned that the _memory_cache dictionary is saving values as 32 bit. I need to look more into how cython implements dictionaries/maps.